File: atcoder.jp/PyPy3 (7.3.0)/abc243/abc243_d.py
N, X = map(int,input().split())
S = list(input())

T = ['x']
for s in S:
    if s == 'U' and T[-1] in 'LR':
        T.pop()
    else:
        T.append(s)

for s in T[1:]:
    if s == 'U':
        X >>= 1
    elif s == 'L':
        X <<= 1
    else:
        X = (X << 1) + 1

print(X)

File: atcoder.jp/Rust (1.42.0)/abc243/abc243_d.rs
use proconio::{fastout, input, marker::Chars};

#[fastout]
fn main() {
    input! {
        _n: usize,
        x: usize,
        s: Chars,
    }

    let s = s.into_iter().fold(Vec::new(), |mut p, c| {
        match c {
            'U' => {
                if vec![Some(&'R'), Some(&'L')].contains(&p.last()) {
                    p.pop();
                } else {
                    p.push(c);
                }
            }
            _ => p.push(c),
        }
        p
    });

    let mut ans = x;
    for c in s {
        ans = match c {
            'U' => ans >> 1,
            'L' => ans << 1,
            'R' => (ans << 1) + 1,
            _ => unreachable!(),
        };
    }
    println!("{}", ans);
}
